File: leetcode/103.binary-tree-zigzag-level-order-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> zigzagLevelOrder(TreeNode* root) {
        if (root == NULL)
        {
            return vector<vector<int>>();
        }

        queue<TreeNode*> q;
        q.push(root);

        vector<vector<int>> res;
        bool reversed = false;

        while (!q.empty())
        {
            vector<int> ans;

            int size = q.size();
            for (auto i = 0; i < size; ++i)
            {
                TreeNode *tmp = q.front();
                q.pop();

                ans.push_back(tmp->val);

                if (tmp->left)
                {
                    q.push(tmp->left);
                }
                
                if (tmp->right)
                {
                    q.push(tmp->right);
                }
            }

            if (reversed)
            {
                reverse(ans.begin(), ans.end());
            }

            res.push_back(ans);
            reversed = !reversed;
        }

        return res;
    }
};

File: leetcode/103.binary-tree-zigzag-level-order-traversal.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def zigzagLevelOrder(self, root):
        """
        :type root: TreeNode
        :rtype: List[List[int]]
        """
        
        if root is None:
            return []

        res = []
        q = [root]
        backward = False

        while len(q) > 0:
            
            ans = []
            size = len(q)
            for i in range(size):
                tmp = q.pop(0)
                ans.append(tmp.val)

                if tmp.left is not None:
                    q.append(tmp.left)

                if tmp.right is not None:
                    q.append(tmp.right)

            if backward:
                ans = ans[::-1]

            res.append(ans)
            backward = not backward

        return res
